src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventPipeMetadataGenerator.cs

```diff
@@ -21,21 +22,37 @@ private enum MetadataTag
         public byte[]? GenerateEventMetadata(EventMetadata eventMetadata)
         {
             ParameterInfo[] parameters = eventMetadata.Parameters;
-            EventParameterInfo[] eventParams = new EventParameterInfo[parameters.Length];
+
+            if (parameters.Length == 0)
+            {
+                return GenerateMetadata(
+                    eventMetadata.Descriptor.EventId,
+                    eventMetadata.Name,
+                    eventMetadata.Descriptor.Keywords,
+                    eventMetadata.Descriptor.Level,
+                    eventMetadata.Descriptor.Version,
+                    (EventOpcode)eventMetadata.Descriptor.Opcode,
+                    default);
+            }
+
+            EventParameterInfo[] array = ArrayPool<EventParameterInfo>.Shared.Rent(parameters.Length);
```

Comment:
Nothing else in the process after this is likely to use these arrays, right?  How much memory does the pool end up rooting for the rest of the process as a result of this?

Reply:
Updated the after to get the other allocs; alas can't `stackalloc` a array with object refs

![image](https://user-images.githubusercontent.com/1142958/110486995-139f3880-80e5-11eb-8b95-bb4f3eab89e4.png)

Reply:
All the allocations are from GenerateEventMetadata; which as discussed in other issues, could be source generated, however is also from ApplicationInsights so would need to be a public generator rather than internal to runtime

![image](https://user-images.githubusercontent.com/1142958/110488561-80670280-80e6-11eb-99fb-46b36402a218.png)

Reply:
>  How much memory does the pool end up rooting for the rest of the process as a result of this?

Creating ArrayPool instance for EventParameterInfo is likely more expensive than 23kB of short-lived GC allocations.

Reply:
> Creating ArrayPool instance for EventParameterInfo is likely more expensive than 23kB of short-lived GC allocations.

Yes, that was what I was hinting at :smile:

Reply:
Looking at distribution of lengths could change it for a short struct with multiple fields and make a span over that

Params | Count |
------:|------:|
0	| 10   |
1	| 74   |
2	| 176  |
3	| 96   |
4	| 32   |
5	| 17   |
6	| 14   |
7	| 2    |
8	| 6    |
9	| 3    |
10	| 4    |
11	| 5    |
12	| 3    |
13	| 2    |
15	| 2    |
16	| 1    |

A 6 field struct covers 93% of lengths

![image](https://user-images.githubusercontent.com/1142958/110533140-96d88280-8115-11eb-87f5-99b0a618f04b.png)